### ISSUE TITLE
Add .has_rsample property to all distributions

### ISSUE THREAD

**Issue by fritzo** [probtorch/pytorch#7]
Probtorch PR #21 
PyTorch PR https://github.com/pytorch/pytorch/pull/3886

**Comment by jwvdm** [comment]
I have a tiny procedural comment: Can we decide on which way we will spell reparameterized?

1. reparametrized (this is what @alicanb has)
2. reparameterized (this is what Tensorflow has)

I don't mind either way, as long as we enforce consistency from here on out :)

**Comment by alicanb** [comment]
Wow, I've never noticed that. Let's go with reparameterized, reparameterization, etc...

**Comment by fritzo** [comment]
Counting google hits:
- `reparametrized` has 180k hits on web, 11k hits on arxiv.org
- `reparameterized` has 67k hits on web, 1.8k hits on arxiv.org

Also, Pyro use pytest which uses the spelling `parametrize`. This always trips me up because Pyro spells it `reparameterized`.

**Comment by fritzo** [comment]
More importantly, I think we should aim to *not* have a reparmeterized property. We should aim for all continuous distributions to have reparameterized implementations (depending on whether params require grads), and no discrete distributions to be reparameterized; thus we could eventually replace `.reparamterized` with `.is_discrete`.

**Comment by tbrx** [comment]
Is there any use case (computational or otherwise) for *not* internally using a reparameterized variant of the sampler when available?

**Comment by fritzo** [comment]
@tbrx I believe the only plausible case is when you're not computing gradients and the reparameterized version is more expensive, e.g. `Gamma` will need to store extra information to compute gradients. I think the cleanest interface for `Distribution.sample()` would be to decide internally based on params (e.g.  `alpha.requires_grad` and `beta.requires_grad` for `Gamma`). This way if a user wants a non-reparameterized version they should call `.detach()` on either the parameters or the resulting sample.

**Comment by jwvdm** [comment]
Regarding: `.reparameterized` vs `.is_discrete`

It seems a strong assumption that all distributions with continuous support will be reparameterizable. For this reason, I would be hesitant to simply replace `.reparameterized` with `.is_discrete`

**Comment by jwvdm** [comment]
Regarding: `requires_grad` and `.detach()`

I think there are two separate use cases here

1. Parameters have `requires_grad=False`. In this case we should clearly also set `requires_grad=False` for the sampled value. 

2. Parameters have `requires_grad=True`, but we still want to draw a sample with `requires_grad=False` (e.g. because we are explicitly calculating a reinforce-style estimate)

The best way to implement 2 might be to allow `Distribution.sample(requires_grad=False)` for reparameterized distributions (which I think somebody might have already suggested in a different issue).

**Comment by fritzo** [comment]
Done as of @alicanb 's https://github.com/pytorch/pytorch/pull/4142
